test/template/test_flextemplate.py

```diff
@@ -508,3 +508,67 @@ def test_flextemplate_leak(tmp_path):  # issue #570
     pdf.ln()
     pdf.cell(text="after", new_x="LEFT", new_y="NEXT")
     assert_pdf_equal(pdf, HERE / "flextemplate_leak.pdf", tmp_path)
+
+
+def test_flextemplate_wrapmode(tmp_path):
```

Comment:
To prevent regressions with possible future changes you should build a similar set of tests for the conventional templates as well. We want as many code paths covered as possible (ideally all of them).

If all of those combinations are supposed to work, then all of them need tests:

* FlexTemplate python dicts
* FlexTemplate CSV
* Template python dicts
* Template CSV

Reply:
Hi @gmischler I've added in tests for both flextemplate and template methods, including using both python dicts and CSV (with an example included in `mycsvfile.csv` that is referenced across multiple tests that I have confirmed pass with new generated references that I've manually inspected to be correct).

I re-installed pre-commit hooks on my computer --- I didn't realise I hadn't fully got that going with my previous commits.   This all works --- with the exception of pylint, where there seems to be some kind of error related to extensions.  If you think that's on my end, I can look into it, but if its a general problem others experience then maybe it doesn't matter?

There was also a warning re duplicate code because the test I made in test_template.py and test_flextemplate.py are basically the same. Not sure if that matters.

I'll paste the pre-commit check below so you can see what you think:
<details>
  <summary>Pre-commit details</summary

```
D:\projects\repos\fpdf2
> git commit -a -m "added in additional tests for csv formatting using wrapmode=CHAR"
CRLF end-lines remover................................................................Passed
Tabs remover..........................................................................Passed
Set file permissions..................................................................Passed
check for merge conflicts.............................................................Passed
check that executables have shebangs..................................................Passed
check that scripts with shebangs are executable.......................................Passed
black.................................................................................Passed
Ensure no print() statement appears in fpdf2 sources..................................Passed
Ensure no generate=True is left in a call to assert_pdf_equal().......................Passed
pylint................................................................................Failed
- hook id: pylint
- exit code: 14

*************
configuration:1:0: E0013: Plugin 'pylint.extensions.dunder' is impossible to load, is it installed ? ('No module named 'pylint.extensions.dunder'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.comparison_placement' is impossible to load, is it installed ? ('No module named 'pylint.extensions.comparison_placement'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.private_import' is impossible to load, is it installed ? ('No module named 'pylint.extensions.private_import'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.redefined_loop_name' is impossible to load, is it installed ? ('No module named 'pylint.extensions.redefined_loop_name'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.consider_refactoring_into_while_condition' is impossible to load, is it installed ? ('No module named 'pylint.extensions.consider_refactoring_into_while_condition'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.no_self_use' is impossible to load, is it installed ? ('No module named 'pylint.extensions.no_self_use'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.for_any_all' is impossible to load, is it installed ? ('No module named 'pylint.extensions.for_any_all'') (bad-plugin-value)
*************
configuration:1:0: E0013: Plugin 'pylint.extensions.eq_without_hash' is impossible to load, is it installed ? ('No module named 'pylint.extensions.eq_without_hash'') (bad-plugin-value)
************* Module fpdf.template
fpdf\template.py:124:24: E1101: Instance of 'tuple' has no '__name__' member (no-member)
************* Module test.template.test_template
test\template\test_template.py:574:0: W1404: Implicit string concatenation found in assignment (implicit-str-concat)
test\template\test_template.py:1:0: R0801: Similar lines in 2 files
==test.template.test_flextemplate:514
==test.template.test_template:595
    elements = [
        {
            "name": "multi",
            "type": "T",
            "x1": 80,
            "y1": 10,
            "x2": 170,
            "y2": 15,
            "text": "If multiline is False, the text should still not wrap even if wrapmode is specified.",
            "background": 0xEEFFFF,
            "multiline": False,
            "wrapmode": "WORD",
        },
        {
            "name": "multi",
            "type": "T",
            "x1": 80,
            "y1": 40,
            "x2": 170,
            "y2": 45,
            "text": "If multiline is True, and wrapmode is omitted, the text should wrap by word and not "
            + "cause an error due to omission of the wrapmode argument.",
            "background": 0xEEFFFF,
            "multiline": True,
        },
        {
            "name": "multi",
            "type": "T",
            "x1": 80,
            "y1": 70,
            "x2": 170,
            "y2": 75,
            "text": "If multiline is True and the wrapmode argument is provided, it should not cause a "
            + "problem even if using the default wrapmode of 'WORD'.",
            "background": 0xEEFFFF,
            "multiline": True,
            "wrapmode": "WORD",
        },
        {
            "name": "multi",
            "type": "T",
            "x1": 80,
            "y1": 100,
            "x2": 170,
            "y2": 105,
            "text": "If multiline is True and the wrapmode argument is provided, it should result in "
            + "wrapping based on characters instead of words, regardless of language (i.e. even though "
            + "this is designed to support scripts like Chinese and Japanese, wrapping long sentences "
            + "like this in English still demonstrates functionality.)",
            "background": 0xEEFFFF,
            "multiline": True,
            "wrapmode": "CHAR",
        },
    ] (duplicate-code)

------------------------------------------------------------------
Your code has been rated at 9.88/10 (previous run: 9.88/10, +0.00)
```

</details>

thanks for the guidance on making this PR!

Reply:
I run black and pylint manually instead of with pre-commit hooks, and have never encountered any errors related to extensions. I don't understand those mechanisms well enough to tell what the problem is on your system. Maybe @Lucas-C has an idea.

However, you should do something about the duplicate code warning. If there is no good way to move the duplicated part to a seperate file (unifying it), then you need to disable that specific warning in that specific place.